File: backend/Vertx3-Kotlin/src/main/java/com/weweb/auth/context/AuthContext.kt
package com.weweb.auth.context

import com.weweb.auth.exception.UnauthenticatedException
import com.weweb.auth.exception.UnauthorizedException

class AuthContext(val user: UserContext, val permissionsMap: Map<String, Array<String>>) {

    fun isAuthenticated(): AuthContext {
        if (user.username.isEmpty()) {
            throw UnauthenticatedException("User needs to be authenticated.")
        }
        return this
    }

    fun hasRole(role: String): AuthContext {
        isAuthenticated()
        if (booleanHasRole(role)) {
            return this;
        }
        throw UnauthorizedException("User [${user.username}] does not have the required roles.")
    }

    fun hasAnyRole(vararg roles: String): AuthContext {
        isAuthenticated()
        for (role in roles) {
            if (booleanHasRole(role)) {
                return this;
            }
        }
        throw UnauthorizedException("User [${user.username}] does not have the required roles.")
    }

    fun hasAllRoles(vararg roles: String): AuthContext {
        isAuthenticated()
        for (role in roles) {
            if (!booleanHasRole(role)) {
                throw UnauthorizedException("User [${user.username}] does not have the required roles.")
            }
        }
        return this
    }

    fun hasPermission(permission: String): AuthContext {
        isAuthenticated()
        if (booleanHasPermission(permission)) {
            return this
        }
        throw UnauthorizedException("User [${user.username}] does not have the required permissions.")
    }

    fun hasAnyPermission(vararg permissions: String): AuthContext {
        isAuthenticated()
        for (permission in permissions) {
            if (booleanHasPermission(permission)) {
                return this
            }
        }
        throw UnauthorizedException("User [${user.username}] does not have the required permissions.")
    }

    fun hasAllPermissions(vararg permissions: String): AuthContext {
        isAuthenticated()
        for (permission in permissions) {
            if (!booleanHasPermission(permission)) {
                throw UnauthorizedException("User [${user.username}] does not have the required permissions.")
            }
        }
        return this
    }

    private fun booleanHasRole(role: String): Boolean {
        for (userRole in user.roles) {
            if (userRole.equals(role))
                return true
        }
        return false
    }

    private fun booleanHasPermission(permission: String): Boolean {
        val permissionRoles = permissionsMap[permission]
        if (permissionRoles != null) {
            for (permissionRole in permissionRoles) {
                if (booleanHasRole(permissionRole)) {
                    return true
                }
            }
        }
        return false
    }

}

File: backend/Go-Chi/src/auth/service/authContext.go
package service

import "github.com/ufoscout/WeWeb/backend/Go-Chi/src/auth/model"

type AuthContext interface {
	isAuthenticated() bool
	hasRole(roleName string) bool
	hasAnyRole(roleNames ...string) bool
	hasAllRoles(roleNames ...string) bool
}

func NewAuthContext(auth model.Auth) AuthContext {
	return &authContextImpl{
		Auth: auth,
	}
}

type authContextImpl struct {
	Auth model.Auth;
}

func (ac *authContextImpl) isAuthenticated() bool {
	return ac.Auth.Username != ""
}

func (ac *authContextImpl) hasRole(roleName string) bool {
	return contains(ac.Auth.Roles, roleName)
}

func (ac *authContextImpl) hasAnyRole(roleNames ...string) bool {
	return true
}

func (ac *authContextImpl) hasAllRoles(roleNames ...string) bool {
	return true
}


func contains(s []string, e string) bool {
	for _, a := range s {
		if a == e {
			return true
		}
	}
	return false
}
